File: code/sorting/src/sleep_sort/sleep_sort.js
/**
 * Sleep Sort Implementation in JavaScript
 *
 * @author  Ahmar Siddiqui <ahmar.siddiqui@gmail.com>
 * @github @ahhmarr
 * @date   07/Oct/2017
 * Part of Cosmos by OpenGenus Foundation
 */

function sleepNumber(number) {
  // The timeout is (n^2) so that close numbers are no longer close.
  // Example: 1 is really close to 2, but 1^2 isn't close to 2^2.
  setTimeout(function() {
    console.log(number);
  }, number * number);
}

function sleepSort(array) {
  for (i = 0; i < array.length; ++i) {
    sleepNumber(array[i]);
  }
}

// Generate 100 random numbers for sorting.
randoms = [];
for (var i = 0; i < 100; ++i) {
  // Random number between 1 - 100
  randomInt = Math.floor(Math.random() * 100 + 1);
  randoms.push(randomInt);
}

// Sort said random numbers.
sleepSort(randoms);

File: cosmos-master/code/sorting/Sleep_Sort/Sleep_Sort.go
/* Part of Cosmos by OpenGenus Foundation */
package main

import "fmt"
import "runtime"
import "time"

func sleepNumber(number int, results chan<- int) {
	time.Sleep(time.Duration(number) * time.Second)
	results <- number
}

func sleepSort(data []int) {
	results := make(chan int, len(data))
	for _, v := range data {
		go sleepNumber(v, results)
	}

	for i, _ := range data {
		data[i] = <-results
	}
	close(results)
}

func main() {
	data := []int{8, 2, 4, 5, 6, 7, 1, 4, 10}
	runtime.GOMAXPROCS(len(data))
	fmt.Println("Before sorting", data)
	sleepSort(data)
	fmt.Println("After sorting", data)
}
